### PR TITLE
Add option to restrict vtable function pointers

### LINKED ISSUE #521: Use vtable cache to restrict CBMC function pointers
Try using the RMC vtable cache to restrict CBMC function pointers via this intertface: http://cprover.diffblue.com/md__home_travis_build_diffblue_cbmc_doc_architectural_restrict-function-pointer.html

### PR DESCRIPTION
**Pull request by avanhatt**
### Description of changes: 

RMC can currently take a long time to verify code with lots of virtual function pointer calls (or, never terminate). This is in part because CBMC uses a very general heuristic to reason about function pointers: it assumes any function of the right type signature could be a target. CBMC does allow you to restrict possible function targets via a command line option and auxiliary json file: http://cprover.diffblue.com/md__home_travis_build_diffblue_cbmc_doc_architectural_restrict-function-pointer.html

Note, using this functionality requires this branch of CBMC, https://github.com/diffblue/cbmc/pull/6376, tracking issue to update RMC once that merges: https://github.com/model-checking/rmc/issues/671

For virtual function pointer through a vtable, we can often do much better than CBMC's default heuristic. In particular, we can virtually guarantee (unless the user transmutes the vtable memory) that a call through a vtable will be one of the trait-defined method for that trait type that we have previously seen in codegen. Because CBMC asserts(false) if the function pointer is not actually within the restricted set, this optimization is sound.

This is especially helpful for `drop_in_place` destructor calls through a vtable. Many functions have types that match `drop` since it takes a single self argument and returns unit. In a Firecracker example, we have seen 300+ function targets identified by the default CBMC heuristic for a single virtual drop call, which otherwise renders CBMC verification basically intractable (think: a massive, 300+ arm `if` block after CBMC performs function pointer elimination). 

This PR add a `VtableCtx` object that build a map from virtual call sites to possible trait-defined method implementations, then writes out the restrictions as a JSON file. 

You can invoke this on single Rust files by passing a `--restrict-vtables` flag to `rmc`, or on full crates with dependencies by adding the following to your harness script (which should eventually be handled by `cargo rmc`:
```
RUST_BACKTRACE=full RUSTFLAGS=" -Z restrict_vtable_fn_ptrs ...
... 
rmc-link-restrictions . <output dir>
...
goto-instrument  --function-pointer-restrictions-file restrictions.json ... 

```

Example:

Say we have two traits that both implement a `noise` method, `Animal` and `Other`. Without the vtable restriction flag, any call to `noise` _and_ any implicit `drop` on a dynamic trait object would unroll to every implementation of either trait:

```rust
  if(!(var_21.vtable->3 == (int (*)(void *))Sheep::noise))
  {
    if(var_21.vtable->3 == (int (*)(void *))i32::noise)
      goto __CPROVER_DUMP_L11;

    if(var_21.vtable->3 == (int (*)(void *))Cow::noise)
      goto __CPROVER_DUMP_L12;

    __CPROVER_assume(0);
  }
```

Now, we see each call site only considers the implementations that match that trait type:
```rust
// Animal::noise 
  if(!(main_1_function_pointer_call_1 == (int (*)(void *))Sheep::noise))
  {
    if(main_1_function_pointer_call_1 == (int (*)(void *))Cow::noise)
      goto __CPROVER_DUMP_L2;

    /* dereferenced function pointer at must be one of [Sheep::noise, Cow::noise] */
    assert(0);
    __CPROVER_assume(0);
  }
  
// Other::noise
  if(!(main_2_function_pointer_call_1 == (int (*)(void *))i32::noise))
  {
    /* dereferenced function pointer at must be i32::noise */
    assert(0);
    __CPROVER_assume(0);
  }
``` 

We also get the benefit of CBMC inserting an `assert(0)` at the end of the block instead of just an `assume(0)`. 

### Resolved issues:

Resolves #521

### Call-outs:

- Using this feature will require this CBMC update to land (which fixes a limitation with command line restriction validation): https://github.com/diffblue/cbmc/pull/6376.

### Testing:

* How is this change tested?

- Manually improved performance on the Firecracker block device example.

* Is this a refactor change?

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Review by celinval — CHANGES_REQUESTED** [review]
Is it possible to make this change less CBMC dependent? Split between collecting potential candidates and doing the CBMC translation after that.

**Comment by avanhatt** [comment]
@celinval Agreed on the goal. I tried to keep the totally CBMC-specific logic to within `get_virtual_function_restrictions`, do you see other place that it would make sense to split off?

**Comment by celinval** [comment]
@avanhatt I was thinking about whether we should split this file: compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs

**Comment by avanhatt** [comment]
@celinval many rebases and switching to serde later, this is now ready for re-review! The CBMC specifics are now in an entirely different crate.

File: compiler/rustc_codegen_rmc/src/codegen/rvalue.rs
Comment (danielsn) [inline compiler/rustc_codegen_rmc/src/codegen/rvalue.rs:725]: This clone is going to be super expensive. Please clone after the lookup!

File: compiler/rustc_codegen_rmc/src/codegen/rvalue.rs
Comment (danielsn) [inline compiler/rustc_codegen_rmc/src/codegen/rvalue.rs:782]: Same here

File: compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs
Comment (danielsn) [inline compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs:33]: Should this be internedstring?

File: compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs
Comment (danielsn) [inline compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs:142]: Is the tostring needed

File: library/rmc_restrictions/src/lib.rs
Comment (danielsn) [inline library/rmc_restrictions/src/lib.rs:12]: Interned string

File: library/rmc_restrictions/src/lib.rs
Comment (danielsn) [inline library/rmc_restrictions/src/lib.rs:21]: What is the format of this

File: src/tools/rmc-link-restrictions/src/main.rs
Comment (danielsn) [inline src/tools/rmc-link-restrictions/src/main.rs:17]: Is this the right place to put the clone?

**Review by danielsn — CHANGES_REQUESTED** [review]

File: scripts/cargo-rmc
Comment (danielsn) [inline scripts/cargo-rmc:68]: `.restrictions.json`

File: library/rmc_restrictions/src/lib.rs
Comment (danielsn) [inline library/rmc_restrictions/src/lib.rs:27]: InternedString

File: compiler/rustc_codegen_rmc/src/codegen/assumptions.rs
Comment (avanhatt) [reply compiler/rustc_codegen_rmc/src/codegen/assumptions.rs:182]: we talked about this in a chime sync, this is just the set of assumptions on the type, so returning `None` is no less safe than `unimplemented!` but let's us not crash.

File: src/test/rmc/DynTrait/vtable_restrictions_fail_fixme.rs
Comment (avanhatt) [reply src/test/rmc/DynTrait/vtable_restrictions_fail_fixme.rs:74]: No, but there is no safe/stable way to mangle the vtable AFAIK, and the test would fail if it were to change.

File: compiler/rustc_codegen_rmc/src/context/goto_ctx.rs
Comment (avanhatt) [reply compiler/rustc_codegen_rmc/src/context/goto_ctx.rs:64]: do you like `emit_vtable_restrictions` better?

Comment (danielsn) [reply compiler/rustc_codegen_rmc/src/context/goto_ctx.rs:64]: Yeah

File: scripts/rmc.py
Comment (danielsn) [inline scripts/rmc.py:62]: Why the new whitespace?

Comment (avanhatt) [reply scripts/rmc.py:62]: We check python formatting with `autopep8` on changed files, it inserted the whitepace

Comment (avanhatt) [reply scripts/rmc.py:62]: Ah didn't realize quite how copious it was, might need to move some of these to doc comments to make it happy

File: compiler/rustc_session/src/options.rs
Comment (tedinski) [inline compiler/rustc_session/src/options.rs:1288]: Is there a strong reason to even make this a flag? Can we emit these files unconditionally?

Comment (avanhatt) [reply compiler/rustc_session/src/options.rs:1288]: The main reasons are that currently emitting restrictions currently (1) requires wrapper functions at ever vtable call, because CBMC uses fragile naming conventions for specifying the function call site, and (2) the restrictions could be incomplete if the Rust code is doing certain types of casts that we don't handle, meaning the restriction set would be too small.

The 3rd even more transient reason is that the CBMC PR has not landed, so restrictions fail on our minimum required CBMC version.

File: compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs
Comment (avanhatt) [reply compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs:117]: Added!

File: compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs
Comment (avanhatt) [reply compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs:164]: Unless something is wrong elsewhere, but added an unreachable assertion.

File: src/tools/rmc-link-restrictions/src/main.rs
Comment (avanhatt) [reply src/tools/rmc-link-restrictions/src/main.rs:10]: it's a single call to `serde_json::to_value(&output).unwrap()` so I don't feel strongly about that, but happy to move if you prefer?

File: src/tools/rmc-link-restrictions/src/main.rs
Comment (avanhatt) [reply src/tools/rmc-link-restrictions/src/main.rs:54]: I suspect this is good enough for now, given how our cargo scripts work.

File: compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs
Comment (avanhatt) [reply compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs]: Switched to named value from MIR.

File: compiler/rustc_codegen_rmc/src/codegen/rvalue.rs
Comment (avanhatt) [reply compiler/rustc_codegen_rmc/src/codegen/rvalue.rs]: Totally just a mistake, thanks!

File: compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs
Comment (avanhatt) [reply compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs]: Mistake.

File: compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs
Comment (avanhatt) [reply compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs]: Not anymore at least

File: library/rmc_restrictions/src/lib.rs
Comment (avanhatt) [reply library/rmc_restrictions/src/lib.rs]: done

File: library/rmc_restrictions/src/lib.rs
Comment (avanhatt) [reply library/rmc_restrictions/src/lib.rs]: It's just the name now, renamed the variable to make clear

File: src/tools/rmc-link-restrictions/src/main.rs
Comment (avanhatt) [reply src/tools/rmc-link-restrictions/src/main.rs]: moved

File: compiler/cbmc/src/irep/serialize.rs
Comment (danielsn) [inline compiler/cbmc/src/irep/serialize.rs:101]: This should really go with InternedString itself.

File: Cargo.lock
Comment (danielsn) [inline Cargo.lock:3350]: What are we using from here?

File: compiler/rustc_codegen_rmc/src/codegen/rvalue.rs
Comment (danielsn) [inline compiler/rustc_codegen_rmc/src/codegen/rvalue.rs:782]: using `ensure` would be more canonical here

File: compiler/rustc_codegen_rmc/src/compiler_interface.rs
Comment (danielsn) [inline compiler/rustc_codegen_rmc/src/compiler_interface.rs:134]: Why do we need to do this, instead of just keeping the `<String, type>` map?

File: compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs
Comment (danielsn) [inline compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs:45]: why not just make this the same name as the struct field?

File: compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs
Comment (danielsn) [inline compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs:66]: clone not needed here. 
```Rust
        let key = TraitDefinedMethod { trait_name, vtable_idx: method };
```

File: compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs
Comment (danielsn) [inline compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs:99]: don't need to say `trait_name: trait_name`, `trait_name` would do

File: compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs
Comment (danielsn) [inline compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs:138]: Is there any risk here if MIR changes?

**Review by danielsn — APPROVED** [review]
Approved with comments

File: src/tools/rmc-link-restrictions/src/main.rs
Comment (danielsn) [reply src/tools/rmc-link-restrictions/src/main.rs:10]: It just feels weird given the function name

File: compiler/rustc_session/src/options.rs
Comment (tedinski) [reply compiler/rustc_session/src/options.rs]: Could we get that broken into an issue, with e.g. links to CBMC issues or PRs we depend on?

This change is awesome, and I'd like to make sure we have a well-lit path to making it the always-on default. e.g. do we know what the "certain types of casts that we don't handle" are, and can we document that?

(This is obviously not a blocker to merging, I just don't want it accidentally dropped or anything!)

File: compiler/rustc_codegen_rmc/src/compiler_interface.rs
Comment (avanhatt) [reply compiler/rustc_codegen_rmc/src/compiler_interface.rs:134]: Presumably because `TyS` doesn't have a serialization, the only thing I'm touching here is changing the string to be interned so I would rather not change more.

File: compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs
Comment (avanhatt) [reply compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs:138]: Less so now that we have the `type_map`, we are normalizing the MIR on both sides instead of relying on string editing.

File: Cargo.lock
Comment (avanhatt) [reply Cargo.lock:3350]: The compiler's `FxHashMap`, not strictly necessary/could be replaced with a vanilla `HashMap`

Comment (danielsn) [reply Cargo.lock:3350]: If easy, could do that here; if not, lets just do that as a cleanup PR

File: compiler/rustc_codegen_rmc/src/codegen/rvalue.rs
Comment (danielsn) [inline compiler/rustc_codegen_rmc/src/codegen/rvalue.rs:793]: Not for this PR, but this feels like something that would happen often enough to be worth a helper function

File: compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs
Comment (danielsn) [inline compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs:69]: Is there any reason we might want to keep this list ordered?

File: compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs
Comment (danielsn) [inline compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs:90]: Where does it increment the index??

File: compiler/rustc_codegen_rmc/src/compiler_interface.rs
Comment (danielsn) [reply compiler/rustc_codegen_rmc/src/compiler_interface.rs:134]: OK

File: compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs
Comment (avanhatt) [reply compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs]: This comment was stale, removed.

File: compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs
Comment (avanhatt) [reply compiler/rustc_codegen_rmc/src/context/vtable_ctx.rs:69]: Not currently.

**Review by celinval — COMMENTED** [review]
I haven't finished reviewing this but so far I haven't seen anything critical. Once @danielsn approves this, I'm ok with merging and making the changes as a follow up PR.

Thanks

**Review by celinval — APPROVED** [review]

**Comment by avanhatt** [comment]
Will address followups in https://github.com/model-checking/rmc/issues/671, merging to unblock Ted's linked PR.

**Pull request merged**
